src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -260,6 +248,28 @@ private IISAwaitable DoFlushAsync()
 
         public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationToken = default(CancellationToken))
         {
+            // To avoid concurrent reading and writing, if we have a pending read,
+            // we must cancel it. TODO check if we need to lock here. I don't think we do because cancel_io will
+            // noop if there is no data to write, and OnAsyncCompletion should only be triggered once per read/write.
+            if (!_wasUpgraded)
+            {
+                lock (_readCancelSync)
+                {
+                    switch (_currentOperationType)
+                    {
+                        case CurrentOperationType.Read:
+                            NativeMethods.http_cancel_io(_pInProcessHandler);
+                            break;
+                        case CurrentOperationType.Write:
+                        case CurrentOperationType.Flush:
+                        case CurrentOperationType.None:
+                        default:
+                            _allowedToRead = false;
+                            break;
+                    }
+                }
+            }
+
             if (!HasResponseStarted)
             {
                 return WriteAsyncAwaited(data, cancellationToken);
```

Comment:
Off topic: Why call an "Awaited" method here already if you don't know yet whether there are any true-async onstarting callbacks registered? I'm guessing there usually aren't such callbacks making the extra async state machine an unnecessary cost. You can look at Kestrel's FireOnStartingMayAwait for how to do this more efficiently.